### PR TITLE
Autostart Linux app with main window hidden

### PR DESCRIPTION
**Pull request by chromakode**
With this change, we will no longer auto-show the main window on Linux desktop startup. If you run `run_keybase` manually from the command-line (as per our [install instructions](https://keybase.io/docs/the_app/install_linux)) the main window *will* show.

Thanks to [Greg Taylor](https://twitter.com/gctaylor/status/760531020437258240) for feedback on this behavior. :fish:

This depends on the `KEYBASE_START_UI` flag @gabriel adds in https://github.com/keybase/client/pull/3695/commits/46a0f62c633f67340c4a9ee931e47e9cd00a0816

:eyeglasses: @keybase/react-hackers

### CODE REVIEW

**Comment by cjb** [comment]
Cc @oconnor663

**Comment by oconnor663** [comment]
Unfortunately I've kind of painted myself into a corner with the logic that the client uses to write that file. If you look at line 49, we skip writing it if the file already exists. The idea there was to make it possible for the user to comment out the exec line to disable autostarting. (Because simply deleting the file is indistinguishable from it never having been installed.) But that also means that clients won't update it when we change this string.

Maybe the right thing to do is to always have the client overwrite it. Then we could provide a "disable autostart" config somewhere in the GUI (or just from the command line to start) that would _both_ delete the existing autostart file _and_ write a setting in one of the usual places to stop us from recreating it.

If we wanted to be real quick and dirty, we could implement something like:
- check for the file `~/.config/autostart/something_something_disable_keybase.something`, and short circuit if it exists
- otherwise always overwrite our autostart desktop file

We could then update the comment in the file we're writing out, to tell people what we're doing and how to create the file that disables it.

**Comment by chromakode** [comment]
Oh this is tricky.

Ideal would be to transition this in some way that respsects existing users' choices, but I can't think of a way to do that which wouldn't require adding the complexity of a full .desktop file parser. We could do some halfway approach like checking for the existence of `#\s*Exec` in the file, but we have no idea what other ways someone could have modified the file.

Side note: while looking at the [autostart spec](https://specifications.freedesktop.org/autostart-spec/autostart-spec-latest.html) I discovered that you can add `Hidden=true` to disable the entry. Might be nice to organize the disable instructions like "To disable, uncomment this line:" instead.

What if we checked if the file contains `#\s*Exec` or `Hidden=true`, and if it doesn't, we overwrote it? This isn't perfect but it'd attempt to respect existing user preference and give us an update path forward. If the user manually *re-enabled* autostart they might end up with an out-of-date `Exec` line on their first run. I'm not as concerned about this, because it seems really improbable and we'd update the file after the first run. (We could also note in comment that to re-enable autostart the user should delete the file so that it is recreated on run)

**Comment by oconnor663** [comment]
One big hammer we could consider is to hold onto the old string, and compare the existing desktop file to that. If it's different at all, don't touch it. Otherwise, go ahead and pave over it with the new stuff. No parsing necessary :)

That said, I'd be ok with just paving over it in all cases if we didn't want to deal with the complexity. It's beta software, hopefully just a one time inconvenience, etc. Whichever approach you prefer.

**Comment by chromakode** [comment]
Alright @oconnor663, this implements the approach we discussed. Tested on Ubuntu with `gnome-session-properties` and it appears to behave as intended.

File: go/install/install_unix.go [hunk 1]
```diff
@@ -45,10 +51,19 @@ func autostartFilePath(context Context) string {
...
-	// If the desktop file already exists, short circuit.
-	if _, err := os.Stat(autostartFilePath(context)); err == nil {
-		return false, nil
+	// If the desktop file already exists and has been disabled by the user, short circuit.
+	if file, err := os.Open(autostartFilePath(context)); err == nil {
+		defer file.Close()
+		// Check if the user has disabled autostart manually.
+		// If the user inserts Hidden=true, our autostart will be ignored.
+		// GNOME inserts X-GNOME-Autostart-enabled=false when startup apps are disabled via gnome-session-properties.
+		autostartDisabledRegex := regexp.MustCompile(`(?m)^Hidden=true|^X-GNOME-Autostart-enabled=false`)
```
Comment (oconnor663) [inline go/install/install_unix.go]: I can never remember precedence rules in regexes. Are we sure `|` binds last? :)

**Comment by oconnor663** [comment]
LGTM! (We talked about adding support for commenting out the Exec line too.)

**Comment by chromakode** [comment]
Ok tested the Exec line check as well w/ a commented-out config from the current version. Looks good.

**Pull request merged**